frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1446 @@
+					// 2. this is the root of a child key, and we start processing child keys (and
+					// should call `migrate_child`).
+					// 3. this is the root of a child key, and we are finishing all child-keys (and
+					// should call `migrate_top`).
+
+					// NOTE: this block is written intentionally to verbosely for easy of
+					// verification.
+					match (
+						top_key.starts_with(DEFAULT_CHILD_STORAGE_KEY_PREFIX),
+						self.prev_tick_child,
+					) {
+						(true, true) => {
+							// we're done with migrating a child-root.
+							self.prev_tick_child = false;
+							self.migrate_top();
+						},
+						(true, false) => {
+							// start going into a child key.
+							let maybe_first_child_key = {
+								let child_top_key = Pallet::<T>::child_io_key(top_key);
+								sp_io::default_child_storage::next_key(child_top_key, &vec![])
+							};
+							if let Some(first_child_key) = maybe_first_child_key {
+								self.current_child = Some(first_child_key);
+								self.prev_tick_child = true;
+								self.migrate_child();
+							} else {
+								self.migrate_top();
+							}
+						},
+						(false, true) => {
+							// should never happen.
+							log!(error, "LOGIC ERROR: unreachable code [0].");
+							Pallet::<T>::halt();
+						},
+						(false, false) => {
+							// continue the top key migration
+							self.migrate_top();
+						},
+					};
+				},
+				(None, Some(_)) => {
+					log!(error, "LOGIC ERROR: unreachable code [1].");
+					Pallet::<T>::halt()
+				},
+				(None, None) => {
+					// nada
+				},
+			}
+		}
+
+		/// Migrate the current child key, setting it to its new value, if one exists.
+		///
+		/// It updates the dynamic counters.
+		fn migrate_child(&mut self) {
+			let current_child =
+				self.current_child.clone().expect("value checked to be `Some`; qed");
+			let current_top = self.current_top.clone().expect("value checked to be `Some`; qed");
+
+			let child_root = Pallet::<T>::child_io_key(&current_top);
+			if let Some(data) = sp_io::default_child_storage::get(child_root, &current_child) {
+				self.dyn_size = self.dyn_size.saturating_add(data.len() as u32);
+				sp_io::default_child_storage::set(child_root, &current_child, &data)
+			}
+			self.dyn_child_items.saturating_inc();
+			let next_key = sp_io::default_child_storage::next_key(child_root, &current_child);
+			self.current_child = next_key;
+			log!(
+				trace,
+				"migrating child key {:?}, next task: {:?}",
+				sp_core::hexdisplay::HexDisplay::from(&current_child),
+				self
+			);
+		}
+
+		/// Migrate the current top key, setting it to its new value, if one exists.
+		///
+		/// It updates the dynamic counters.
+		fn migrate_top(&mut self) {
+			let current_top = self.current_top.clone().expect("value checked to be `Some`; qed");
+			let added_size = if let Some(data) = sp_io::storage::get(&current_top) {
+				self.dyn_size = self.dyn_size.saturating_add(data.len() as u32);
+				sp_io::storage::set(&current_top, &data);
```

Comment:
what's the reason? and what should I do in `else`?

Reply:
The new trie nodes are only use when value are bigger than 32 bytes, otherwhise it is a noop. So for else, simply ignore the write, just be sure to count it as an item and count its size, as it will end up in the pov.

Reply:
still unclear to me why we need this. We definitely want to have bookkeeping for that key, regardless of it having any sizes, which we already do. If the `::set()` is a noop in some cases, let it be, I don't see any harm in it.

Reply:
> We definitely want to have bookkeeping for that key

yes, the idea would be to just skip the `::set()` call (replace all ::set by if size > 32 { ::set }), and keep everything else as it is.
This way we save a bit of memory and processing.
There is indeed no harm in keeping `::set()` just could be optimized.